syntax = "proto3";
package protobufs;

// Message from client and related types

message ClientToTransferNodeMessage {
  enum MessageType {
    AUTHENTICATE = 0;
    START_UPLOAD = 1;
    UPLOAD_DATA = 2;
    FINISHED = 3;
    ERROR = 4;
  }
  MessageType type = 1;
  AuthenticateData authData = 2;
  StartUploadData startData = 3;
  UploadData uploadData = 4;
  FinishedData finishedData = 5;
  int64 timestamp = 6;
}

message AuthenticateData {
  string key = 1;
}

message StartUploadData {
  string filename = 1;
  string mimetype = 2;
  uint64 size = 3;
}

message UploadData {
  uint64 order = 1;
  uint64 size = 2;
  bytes data = 3;
}

message FinishedData {
  string error = 1;
  bool success = 2;
}

// ----
// Message to client and related types

message TransferNodeToClientMessage {
  enum MessageType {
    AUTH_SUCCESS = 0;
    TRANSFER_CREATED = 1;
    RECIPIENTS = 3;
    PROGRESS = 4;
    FINISHED = 5;
    ERROR = 6;
  }

  MessageType type = 1;
  AuthSuccessData authSuccessData = 2;
  TransferCreatedData transferCreatedData = 3;
  RecipientsData recipientsData = 4;
  ProgressData progressData = 5;
  FinishedData finishedData = 6;
  ErrorData errorData = 7;
  int64 timestamp = 8;
}

message AuthSuccessData {

}

message TransferCreatedData {
  string transferId = 1;
  uint64 chunkSize = 2;
  uint64 requestChunks = 3;
}

message RecipientsData {
  message Recipient {
    string ipv4 = 1;
    string ipv6 = 2;
    string identity = 3;
  }
  repeated Recipient recipients = 1;
}

message ProgressData {
  int64 bytesUploaded = 1;
  uint64 requestChunks = 2;
  uint64 chunkSize = 3;
}

message ErrorData {
  string title = 1;
  string jsonDetails = 2;
  bool fatal = 3;
}
